bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -88,33 +90,41 @@ uint32_t ClientsManager::numberOfRequiredReservedPages() const { return required
 
 // Per client:
 // * calculate offset of reserved page start.
+// * Iterate for max number of requests per client
 // * load corresponding page from state-transfer to scratchPage.
 // * Fill its clientInfo.
-// * remove pending request if loaded reply is newer.
+// * At the end of the loop take the latest loaded reply and remove pending request if latest reply is newer.
 void ClientsManager::loadInfoFromReservedPages() {
   for (auto const& [clientId, clientIdx] : clientIdToIndex_) {
     const uint32_t firstPageId = clientIdx * reservedPagesPerClient_;
 
-    if (!stateTransfer_->loadReservedPage(resPageOffset() + firstPageId, sizeOfReservedPage_, scratchPage_)) continue;
-
-    ClientReplyMsgHeader* replyHeader = (ClientReplyMsgHeader*)scratchPage_;
-    ConcordAssert(replyHeader->msgType == 0 || replyHeader->msgType == MsgCode::ClientReply);
-    ConcordAssert(replyHeader->currentPrimaryId == 0);
-    ConcordAssert(replyHeader->replyLength >= 0);
-    ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
-
-    // YS TBD: Multiple replies for client batching should be sorted by incoming time
-    auto& repliesInfo = indexToClientInfo_.at(clientIdx).repliesInfo;
-    if (repliesInfo.size() >= maxNumOfReqsPerClient_) deleteOldestReply(clientId);
-    const auto& res = repliesInfo.insert_or_assign(replyHeader->reqSeqNum, MinTime);
-    const bool added = res.second;
-    LOG_INFO(CL_MNGR, "Added/updated reply message" << KVLOG(clientId, replyHeader->reqSeqNum, added));
-
+    // load all replies into the map of replies one by one
+    for (int i = 0; i < maxNumOfReqsPerClient_; i++) {
+      if (!stateTransfer_->loadReservedPage(
+              resPageOffset() + firstPageId + i, sizeOfReservedPage_, scratchPage_.data() + i * sizeOfReservedPage_))
+        continue;
+
+      auto replyHeader = (ClientReplyMsgHeader*)(scratchPage_.data() + i * sizeOfReservedPage_);
+      auto replyPtr = std::make_shared<ClientReplyMsg>(myId_, replyHeader->replyLength);
+      ConcordAssert(replyHeader->msgType == 0 || replyHeader->msgType == MsgCode::ClientReply);
+      ConcordAssert(replyHeader->currentPrimaryId == 0);
+      ConcordAssert(replyHeader->replyLength >= 0);
+      ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
+
+      auto& repliesInfo = indexToClientInfo_.at(clientIdx).repliesInfo;
+      if (repliesInfo.size() >= maxNumOfReqsPerClient_) deleteOldestReply(clientId);
+      const auto& res = repliesInfo.insert({replyHeader->reqSeqNum, replyPtr});
```

Comment:
I recall there was a problem when used just insert, that's the reason insert_or_assign was chosen.
Are we dealing correct with the situation when the message exists?

Reply:
Changed to insert_or_assign thanks.